crates/re_arrow_store/src/store_view.rs

```diff
@@ -0,0 +1,254 @@
+
+    /// Workaround for downcasting support, simply return `self`:
+    /// ```ignore
+    /// fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+    ///     self
+    /// }
+    /// ```
+    fn as_any_mut(&mut self) -> &mut dyn std::any::Any;
+
+    /// The core of this trait: get notified of changes happening in one or more [`DataStore`]s.
+    ///
+    /// This will be called automatically by the [`DataStore`] itself if the view has been
+    /// registered: [`DataStore::register_view`].
+    /// Or you might want to feed it [`StoreEvent`]s manually, depending on your use case.
+    ///
+    /// ## Example
+    ///
+    /// ```ignore
+    /// fn on_events(&mut self, events: &[StoreEvent]) {
+    ///     use re_arrow_store::StoreDiffKind;
+    ///     for event in events {
+    ///         match event.kind {
+    ///             StoreDiffKind::Addition => println!("Row added: {}", event.row_id),
+    ///             StoreDiffKind::Deletion => println!("Row removed: {}", event.row_id),
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    fn on_events(&mut self, events: &[StoreEvent]);
+}
+
+/// All registered [`StoreView`]s.
+static VIEWS: once_cell::sync::Lazy<RwLock<Vec<SharedStoreView>>> =
+    once_cell::sync::Lazy::new(|| RwLock::new(Vec::new()));
+
+#[derive(Debug, Clone, Copy)]
+pub struct StoreViewHandle(u32);
+
+impl DataStore {
+    /// Registers a [`StoreView`] so it gets automatically notified when data gets added and/or
+    /// removed to/from a [`DataStore`].
+    ///
+    /// Refer to [`StoreEvent`]'s documentation for more information about these events.
+    ///
+    /// ## Scope
+    ///
+    /// Registered [`StoreView`]s are global scope: they get notified of all events from all
+    /// existing [`DataStore`]s, including [`DataStore`]s created after the view was registered.
+    ///
+    /// Use [`StoreEvent::store_id`] to identify the source of an event.
+    ///
+    /// ## Late registration
+    ///
+    /// Views must be registered before a store gets created to guarantee that no events were
+    /// missed.
+    ///
+    /// [`StoreEvent::event_id`] can be used to identify missing events.
+    ///
+    /// ## Ordering
+    ///
+    /// The order in which registered views are notified is undefined and will likely become
+    /// concurrent in the future.
+    ///
+    /// If you need a specific order across multiple views, embed them into an orchestrating view.
+    //
+    // TODO(cmc): send a compacted snapshot to late registerers for bootstrapping
+    pub fn register_view(view: Box<dyn StoreView>) -> StoreViewHandle {
+        let mut views = VIEWS.write();
+        views.push(RwLock::new(view));
+        StoreViewHandle(views.len() as u32 - 1)
+    }
+
+    /// Passes a reference to the downcasted view to the given callback.
+    ///
+    /// Returns `None` if the view doesn't exist or downcasting failed.
+    pub fn with_view<V: StoreView, T, F: FnMut(&V) -> T>(
```

Comment:
I assume this is because it doesn't return the view, but rather takes a closure that is executed with the view as an argument.  As in "With this view, run this function". That said, I also find `with_` confusing here. Maybe `for_`?

Reply:
I mostly based this on the `thread_local` API from the stdlib: https://doc.rust-lang.org/std/thread/struct.LocalKey.html#examples

Reply:
I'll keep it as is for now as it feels somewhat consistent with std at least... definitely feel free to change it later if you come up with anything better.